File: src/com/leetcode/tree/M144_Binary_Tree_Preorder_Traversal.java
package com.leetcode.tree;

import sun.reflect.generics.tree.Tree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Created by LYuan on 2016/10/8.
 * Given a binary tree, return the preorder traversal of its nodes' values. <二叉树的前序遍历>
 * For example: Given binary tree as below,
 *         1
 *       /  \
 *     2     5
 *    / \     \
 *   3  4      6
 * Preorder traversal should return [1,2,3,4,5,6].
 *
 * Function Signature:
 * public List<Integer> preorder(TreeNode root) {...}
 */
public class M144_Binary_Tree_Preorder_Traversal {
    public static void main(String[] args) {
        TreeNode root = TreeNode.Generator(new int[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11});
        System.out.println(preOrder(root));
        System.out.println(preOrder2(root));
        System.out.println(preOrder3(root));
        System.out.println(preOrder_iterative(root));
        System.out.println(preOrder_iterative2(root));
    }

    /** 二叉树的前、中、后序遍历都是针对root而言的。
     *  <前序>：先访问root
     *  <中序>：先访问左儿子，再访问root
     *  <后序>：先访问左儿子，再访问右儿子，再访问root */

    /** 二叉树不是线性存储结构，因此如果想要线性的串行（而非并行多线程）遍历二叉树，必然会在遍历的过程中缓存一部分节点。
     *  对于<递归方式>的遍历，节点缓存的工作被隐式的通过函数调用栈完成。
     *  对于<迭代方式>的遍历，则必须构造额外的栈或队列来辅助才能完成。*/

    /** 递归解法 1：每次递归都创建新List，返回时用addAll()方法将新List合并入最后的结果中。 */
    // 为什么要用addAll而不是add：就是为了写递归方式直接一个方法省事。
    // 因为一般要递归的话，都需要给递归方法添加一些传入参数，这就与给定的解题方法签名不同了，因此一般都是一个wrapper方法搭配真正的递归方法。
    // 这里为了保持给定的解题方法接口签名不变，就需要每次递归都返回该层高度的树所记录的遍历信息，最后汇总到一起返回。
    static List<Integer> preOrder(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root != null) {
            result.add(root.val);
            result.addAll(preOrder(root.left));
            result.addAll(preOrder(root.right));
        }
        return result;
    }

    /** 递归解法 2：使用成员变量游离于递归方法之外，避免每次递归都创建新List。 */
    static List<Integer> result = new ArrayList<>();
    static List<Integer> preOrder2(TreeNode root) {
        if (root != null) {
            result.add(root.val);
            preOrder2(root.left);
            preOrder2(root.right);
        }
        return result;
    }

    /** 递归解法 3：标准解法。使用Wrapper + 递归方法。 */
    static List<Integer> preOrder3(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        preOrder_Recursive(root, result);
        return result;
    }

    static void preOrder_Recursive(TreeNode node, List<Integer> result) {
        if (node != null) {
            result.add(node.val);
            preOrder_Recursive(node.left, result);
            preOrder_Recursive(node.right, result);
        }
    }

    /** 迭代解法 1：标准解法，使用栈辅助。 */
    // 注意：栈不能接受null元素。因此所有push前都要先做null check。
    static List<Integer> preOrder_iterative(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        if (root == null) return result;
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode current = stack.pop();
            result.add(current.val);    // 前序：访问子节点前先把根输出。
            if (current.right != null) stack.push(current.right);
            if (current.left != null) stack.push(current.left);
        }
        return result;
    }

    /** 迭代解法 2：同样使用栈的另一种写法。 */
    // 这是类似于Inorder和Postorder解法的if...else结构的Preorder版本，但是并没有上面的解法直观。
    static List<Integer> preOrder_iterative2(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode current = root;
        while (current != null || !stack.isEmpty()) {
            if (current != null) {
                result.add(current.val);
                if (current.right != null) stack.push(current.right);
                current = current.left;
            }
            else  current = stack.pop();
        }
        return result;
    }
}

File: src/com/leetcode/tree/M144_Binary_Tree_Preorder_Traversal.cc
#include "../common.h"
#include <stack>

/**
 * Created by Michael on 2019/8/11.
 * 
 * Binary Tree Pre-order Traversal
 *      - Depth-First
 *      - root -> left -> right
 * 
 *         1
 *       /  \
 *     2     5
 *    / \     \
 *   3  4      6
 * 
 * Output as: 1, 2, 3, 4, 5, 6
 * 
 */

// Iterative + Stack 解法一
// 循环的设计：从栈中取节点，访问值，压右儿子，压左儿子。
// 
// 循环前将种子压栈：在循环开始前，把种子（root）压入栈中，循环内部根据情况把节点指针压栈或出栈，循环在栈空后终止。这样可以精简循环条件和循环内部的逻辑的设计。
// 迭代的终止条件：栈空。此时所有的有效节点都已经遍历。
// 压栈前 null-check：总是在压栈前做 null-check，包括在一开始检查 root 是否为空，省去在循环内检查，确保每次迭代开始出栈的元素都是有效的。
std::vector<int> preorder_iter(TreeNode* root) {
  std::vector<int> result;
  std::stack<TreeNode*> stack;
  if (root == nullptr) return result;
  stack.push(root);
  while (!stack.empty()) {
    TreeNode* curr = stack.top();     // 完善的 null-check 可以确保栈内元素一定有效，可以放心出栈。curr 是循环体内部的临时缓存，每次循环记忆都会清空。
    stack.pop();
    result.push_back(curr->val);      // root 1st. 这句话才是每次循环对结果的真正贡献。下面的压栈操作都是给后面循环做的准备而已，功能等效于函数调用栈。
    if (curr->right != nullptr) 
      stack.push(curr->right);        // right 3rd
    if (curr->left != nullptr) 
      stack.push(curr->left);         // left 2nd. 注意这里应该压栈，而不是直接让 curr = curr->left，那样会与循环体开始的出栈操作冲突。
  }
  return result;
}

// Iterative + Stack 解法二
// 循环的设计：如果节点不为空，就访问值，压右儿子，然后跳转至左儿子（至于左儿子是否存在不管）；如果节点为空，就出栈取新节点等下一轮。
//
// 循环前没有将种子压栈，这使得循环体内部不能以出栈开头。需要分情况讨论。
// 迭代的终止条件有两重：当前节点为空且栈也为空时。
std::vector<int> preorder_iter2(TreeNode* root) {
  std::vector<int> result;
  std::stack<TreeNode*> stack;
  while (root != nullptr || !stack.empty()) {
    if (root != nullptr) {
      result.push_back(root->val);
      if (root->right != nullptr) stack.push(root->right);
      root = root->left;
    } else {
      root = stack.top();
      stack.pop();
    }
  }
  return result;
}


// Recursion with global vector (or member variable if define class).
// Time ~ O(n), Space ~ O(n)
std::vector<int> vec;
std::vector<int> preorder_recurr (TreeNode* root) {
  if (root != nullptr) {      // only perform null check for current node, not its siblings.
    vec.push_back(root->val);
    preorder_recurr(root->left);
    preorder_recurr(root->right);
  }
  return vec;
}

// Recursion with helper func.
void preorder_helper(TreeNode* root, std::vector<int>& result) {
  if (root == nullptr) return;
  result.push_back(root->val);
  preorder_helper(root->left, result);
  preorder_helper(root->right, result);
}
std::vector<int> preorder_recurr2 (TreeNode* root) {
  std::vector<int> result;
  preorder_helper(root, result);
  return result;
}

int main() {
  TreeNode* root = new TreeNode(1);
  root->left = new TreeNode(2);
  root->right = new TreeNode(5);
  root->left->left = new TreeNode(3);
  root->left->right = new TreeNode(4);
  root->right->right = new TreeNode(6);
  
  std::vector<int> result = preorder_recurr(root);
  for (auto x : result) 
    std::cout << x << ", ";
  std::cout << std::endl;

  std::vector<int> result2 = preorder_recurr2(root);
  for (auto x : result2) 
    std::cout << x << ", ";
  std::cout << std::endl;

  std::vector<int> result3 = preorder_iter(root);
  for (auto x : result3) 
    std::cout << x << ", ";
  std::cout << std::endl;

  std::vector<int> result4 = preorder_iter2(root);
  for (auto x : result4) 
    std::cout << x << ", ";
  std::cout << std::endl;
}